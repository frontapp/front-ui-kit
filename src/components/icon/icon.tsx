import React, {FC} from 'react';
import styled from 'styled-components';

import CheckmarkCircle from '../../assets/icons/checkmarkCircle.svg';
import Checkmark from '../../assets/icons/checkmarkSelected.svg';
import ChevronDown from '../../assets/icons/chevronDown.svg';
import ChevronLeft from '../../assets/icons/chevronLeft.svg';
import ChevronRight from '../../assets/icons/chevronRight.svg';
import ChevronUp from '../../assets/icons/chevronUp.svg';
import Close from '../../assets/icons/close.svg';
import ExternalLink from '../../assets/icons/externalLink.svg';
import Minus from '../../assets/icons/minus.svg';
import EllipsisHorizontal from '../../assets/icons/moreHorizontal.svg';
import EllipsisVertical from '../../assets/icons/moreVertical.svg';
import Plus from '../../assets/icons/new.svg';
import PlusCircle from '../../assets/icons/plusCircle.svg';
import Search from '../../assets/icons/search.svg';
import {greys} from '../../helpers/colorHelpers';

/*
 * Constants.
 */

const defaultSize = 16;
const defaultColor = greys.shade50;

const importedIcons = {
  CheckmarkCircle,
  Checkmark,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  ChevronUp,
  Close,
  ExternalLink,
  Minus,
  EllipsisHorizontal,
  EllipsisVertical,
  Plus,
  PlusCircle,
  Search
};
export type IconName = keyof typeof importedIcons;
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
export const icons = importedIcons as {[name in IconName]: (props: React.SVGAttributes<SVGElement>) => React.ReactElement};

/*
 * Props.
 */

interface IconProps {
  /** The icon to display. */
  name: IconName;
  /** The color of the icon. */
  color?: string;
  /** Controls the width and height of the icon. */
  size?: number;
  /** If set, no color for the icon will be set. This enables inheriting the parents color. */
  shouldDisableColor?: boolean;
}

/*
 * Style.
 */

interface StyledIconDivProps {
  $size: number;
}

const StyledIconDiv = styled.div<StyledIconDivProps>`
  width: ${p => `${p.$size}px`};
  height: ${p => `${p.$size}px`};
`;

/*
 * Component.
 */

export const Icon: FC<IconProps> = props => {
  const {name, size = defaultSize, color = defaultColor, shouldDisableColor} = props;

  // Pull the icon from the list of available icons.
  // The svg file is converted to a function that is called.
  const icon = icons[name];
  if (!icon)
    return null;

  return (
    <StyledIconDiv $size={size}>
      {icon({
        width: size,
        height: size,
        color: shouldDisableColor ? '' : color,
        preserveAspectRatio: "none"
      })}
    </StyledIconDiv>
  );
};
