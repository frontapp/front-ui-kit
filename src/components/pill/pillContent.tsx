import React, {FC} from 'react';
import styled from 'styled-components';

/*
 * Props.
 */

interface PillContentProps {
  /** Children to render. */
  children: React.ReactNode;
}

/*
 * Style.
 */

const StyledPillContentDiv = styled.div`
  grid-area: content;
`;

/*
 * Component.
 */

export const PillContent: FC<PillContentProps> = ({children}) => (
  <StyledPillContentDiv>{children}</StyledPillContentDiv>
);
