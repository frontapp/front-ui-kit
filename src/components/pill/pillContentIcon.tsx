import React, {FC} from 'react';
import styled, {css} from 'styled-components';

import {renderFirstIconOnly} from '../../helpers/renderHelpers';

/*
 * Props.
 */

interface PillContentIconProps {
  /** Position of the icon. If you want an icon in the center, use PillContent and Icon. */
  position?: 'left' | 'right';
  /** Children to render. */
  children: React.ReactNode;
}

/*
 * Style.
 */

interface StyledContentIconDivProps {
  $position: 'left' | 'right';
}

const StyledContentIconDiv = styled.div<StyledContentIconDivProps>`
  ${p => {
    if (p.$position === 'left')
      return css`
        grid-area: left-content;
        margin-right: 8px;
      `;
    return css`
      grid-area: right-content;
      margin-left: 8px;
    `;
  }}
  
`;

/*
 * Component.
 */

export const PillContentIcon: FC<PillContentIconProps> = ({children, position = 'left'}) => (
  <StyledContentIconDiv $position={position}>{renderFirstIconOnly(children)}</StyledContentIconDiv>
);
