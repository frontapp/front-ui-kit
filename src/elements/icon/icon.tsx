import React, {FC} from 'react';
import styled from 'styled-components';

import Archive from '../../assets/icons/archive.svg';
import ArchiveFilled from '../../assets/icons/archiveFilled.svg';
import Assign from '../../assets/icons/assign.svg';
import AssignFilled from '../../assets/icons/assignFilled.svg';
import AttachmentArchive from '../../assets/icons/attachmentArchive.svg';
import AttachmentCalendar from '../../assets/icons/attachmentCalendar.svg';
import AttachmentCode from '../../assets/icons/attachmentCode.svg';
import AttachmentExcel from '../../assets/icons/attachmentExcel.svg';
import AttachmentGeneric from '../../assets/icons/attachmentGeneric.svg';
import AttachmentImage from '../../assets/icons/attachmentImage.svg';
import AttachmentMusic from '../../assets/icons/attachmentMusic.svg';
import AttachmentPdf from '../../assets/icons/attachmentPdf.svg';
import AttachmentPowerpoint from '../../assets/icons/attachmentPowerpoint.svg';
import AttachmentVideo from '../../assets/icons/attachmentVideo.svg';
import AttachmentWord from '../../assets/icons/attachmentWord.svg';
import Calendar from '../../assets/icons/calendar.svg';
import CalendarFilled from '../../assets/icons/calendarFilled.svg';
import CheckmarkBox from '../../assets/icons/checkmarkBox.svg';
import CheckmarkCircle from '../../assets/icons/checkmarkCircle.svg';
import CheckmarkCircleEmpty from '../../assets/icons/checkmarkCircleEmpty.svg';
import Checkmark from '../../assets/icons/checkmarkSelected.svg';
import ChevronDown from '../../assets/icons/chevronDown.svg';
import ChevronLeft from '../../assets/icons/chevronLeft.svg';
import ChevronRight from '../../assets/icons/chevronRight.svg';
import ChevronUp from '../../assets/icons/chevronUp.svg';
import Close from '../../assets/icons/close.svg';
import Copy from '../../assets/icons/copy.svg';
import Edit from '../../assets/icons/edit.svg';
import EditFilled from '../../assets/icons/editFilled.svg';
import Export from '../../assets/icons/export.svg';
import ExternalLink from '../../assets/icons/externalLink.svg';
import Gear from '../../assets/icons/gear.svg';
import GearFilled from '../../assets/icons/gearFilled.svg';
import Import from '../../assets/icons/import.svg';
import Info from '../../assets/icons/info.svg';
import InfoFilled from '../../assets/icons/infoFilled.svg';
import Minus from '../../assets/icons/minus.svg';
import EllipsisHorizontal from '../../assets/icons/moreHorizontal.svg';
import EllipsisVertical from '../../assets/icons/moreVertical.svg';
import Plus from '../../assets/icons/new.svg';
import NewContact from '../../assets/icons/newContact.svg';
import Participant from '../../assets/icons/participant.svg';
import ParticipantFilled from '../../assets/icons/participantFilled.svg';
import PlusCircle from '../../assets/icons/plusCircle.svg';
import Preferences from '../../assets/icons/preferences.svg';
import Search from '../../assets/icons/search.svg';
import Star from '../../assets/icons/star.svg';
import StarFilled from '../../assets/icons/starFilled.svg';
import Trash from '../../assets/icons/trash.svg';
import TrashFilled from '../../assets/icons/trashFilled.svg';
import WarningFilled from '../../assets/icons/warningFilled.svg';
import {greys} from '../../helpers/colorHelpers';

/*
 * Constants.
 */

const defaultSize = 16;
const defaultColor = greys.shade50;

const importedIcons = {
  Archive,
  ArchiveFilled,
  Assign,
  AssignFilled,
  AttachmentArchive,
  AttachmentCalendar,
  AttachmentCode,
  AttachmentExcel,
  AttachmentGeneric,
  AttachmentImage,
  AttachmentMusic,
  AttachmentPdf,
  AttachmentPowerpoint,
  AttachmentVideo,
  AttachmentWord,
  Calendar,
  CalendarFilled,
  CheckmarkCircle,
  CheckmarkCircleEmpty,
  CheckmarkBox,
  Checkmark,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  ChevronUp,
  Close,
  Copy,
  Edit,
  EditFilled,
  EllipsisHorizontal,
  EllipsisVertical,
  Export,
  ExternalLink,
  Gear,
  GearFilled,
  Import,
  Info,
  InfoFilled,
  Minus,
  NewContact,
  Participant,
  ParticipantFilled,
  Plus,
  PlusCircle,
  Preferences,
  Search,
  Star,
  StarFilled,
  Trash,
  TrashFilled,
  WarningFilled
};
export type IconName = keyof typeof importedIcons;
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
export const icons = importedIcons as {
  [name in IconName]: (props: React.SVGAttributes<SVGElement>) => React.ReactElement;
};

/*
 * Props.
 */

interface IconProps {
  /** The icon to display. */
  name: IconName;
  /** The color of the icon. */
  color?: string;
  /** Controls the width and height of the icon. */
  size?: number;
  /** If set, no color for the icon will be set. This enables inheriting the parents color. */
  shouldDisableColor?: boolean;
}

/*
 * Style.
 */

interface StyledIconDivProps {
  $size: number;
}

const StyledIconDiv = styled.div<StyledIconDivProps>`
  width: ${p => `${p.$size}px`};
  height: ${p => `${p.$size}px`};
`;

/*
 * Component.
 */

export const Icon: FC<IconProps> = props => {
  const {name, size = defaultSize, color = defaultColor, shouldDisableColor} = props;

  // Pull the icon from the list of available icons.
  // The svg file is converted to a function that is called.
  const icon = icons[name];
  if (!icon)
    return null;

  return (
    <StyledIconDiv $size={size}>
      {icon({
        name,
        width: size,
        height: size,
        color: shouldDisableColor ? '' : color,
        preserveAspectRatio: 'none'
      })}
    </StyledIconDiv>
  );
};
