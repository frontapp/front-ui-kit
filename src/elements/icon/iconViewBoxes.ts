// Generated metadata for original SVG viewBox values
// Only stores exceptions - most icons use the default "0 0 16 16"
// This ensures we can preserve the original viewBox from SVG files
// when no viewBox prop is explicitly provided

import type { IconName } from './icon';

const ICON_VIEW_BOX_EXCEPTIONS: Partial<Record<IconName, string>> = {
  Export: '0 0 16 17',
  Info20: '0 0 20 20',
  CaretExpand: '0 0 20 20',
  EditSquare: '0 0 20 20',
  GearRounded: '0 0 18 20'
};

const DEFAULT_VIEW_BOX = '0 0 16 16';

/**
 * Gets the original viewBox for an icon name.
 * Returns the exception viewBox if one exists, otherwise returns the default.
 */
export function getIconViewBox(iconName: IconName): string {
  return ICON_VIEW_BOX_EXCEPTIONS[iconName] || DEFAULT_VIEW_BOX;
}
