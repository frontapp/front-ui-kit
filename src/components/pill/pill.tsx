import React, {FC, MouseEventHandler} from 'react';
import styled, {css} from 'styled-components';

import {alphas, greys} from '../../helpers/colorHelpers';
import {fonts, fontSizes} from '../../helpers/fontHelpers';
import {isComponentInChildren} from '../../helpers/renderHelpers';
import {getBackgroundColorFromStyles, getTextColorFromStyles, SelectableComponentColors} from '../../helpers/styleHelpers';
import {PillContent} from './pillContent';

/*
 * Constants.
 */

const defaultPillColors: SelectableComponentColors = {
  textColor: greys.shade90,
  selectedTextColor: greys.white,
  backgroundColor: alphas.black30,
  hoverBackgroundColor: alphas.black40,
  selectedBackgroundColor: greys.shade70
};

/*
 * Props.
 */

interface PillProps {
  /** Whether we should render the selected variant of the colors. */
  isSelected?: boolean;
  /** Custom colors for different states. Note, disabled is not supported. Default is a grey variant. */
  colors?: SelectableComponentColors;
  /** Content to render for the pill. */
  children: React.ReactNode;
  /** Called when the pill is clicked. */
  onClick?: MouseEventHandler;
}

/*
 * Style.
 */

interface StyledPillDivProps {
  $colors: SelectableComponentColors;
  $isSelected?: boolean;
}

const StyledPillDiv = styled.div<StyledPillDivProps>`
  padding: 3px 6px;
  border-radius: 100px;
  border-radius: 4px;
  display: inline-grid;
  grid-template-areas: "left-content content right-content";
  font-family: ${fonts.system};
  box-sizing: border-box;
  font-size: ${fontSizes.medium};
  user-select: none;
  line-height: 16px;

  ${p => addPillStyles(p.$colors, p.$isSelected)}
`;

function addPillStyles(colors: SelectableComponentColors, isSelected?: boolean) {
  if (isSelected)
    return css`
      color: ${getTextColorFromStyles(colors, 'selected')};
      background-color: ${getBackgroundColorFromStyles(colors, 'selected')};
    `;
  return css`
    color: ${getTextColorFromStyles(colors)};
    background-color: ${getBackgroundColorFromStyles(colors)};

    &:hover {
      color: ${getTextColorFromStyles(colors, 'hover')};
      background-color: ${getBackgroundColorFromStyles(colors, 'hover')};
    }
  `;
}

/*
 * Component.
 */

export const Pill: FC<PillProps> = props => {
  const {children, colors = defaultPillColors, isSelected, onClick} = props;

  return (
    <StyledPillDiv $colors={colors} $isSelected={isSelected} onClick={onClick}>
      {renderPillChildren(children)}
    </StyledPillDiv>
  );
};

/*
 * Helpers.
 */

/** Renders the children and checks if we need to wrap the passed in children in the PillContent. */
function renderPillChildren(children: React.ReactNode) {
  const shouldWrapInContent = !isComponentInChildren(children, 'PillContent');
  if (!shouldWrapInContent)
    return children;
  return <PillContent>{children}</PillContent>;
}
