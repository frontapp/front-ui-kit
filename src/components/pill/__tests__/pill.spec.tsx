import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {Icon} from '../../../elements/icon/icon';
import {Pill} from '../pill';
import {PillContent} from '../pillContent';
import {PillContentIcon} from '../pillContentIcon';

describe('<Pill /> Spec', () => {
  describe('Snapshots', () => {
    it('should render using defaults', () => {
      const view = render(<Pill>Pill</Pill>);
      expect(view.container).toMatchSnapshot();
    });

    it('should render selected state', () => {
      const view = render(<Pill isSelected>Pill</Pill>);
      expect(view.container).toMatchSnapshot();
    });

    it('should support rendering a pill with icon', () => {
      const view = render(
        <Pill>
          <PillContent>Pill</PillContent>
          <PillContentIcon>
            <Icon name="Checkmark" />
          </PillContentIcon>
        </Pill>
      );
      expect(view.container).toMatchSnapshot();
    });
  });

  it('should issue onClick events when clicked', async () => {
    const user = userEvent.setup();
    const onClickMock = jest.fn();
    const view = render(<Pill onClick={onClickMock}>Pill</Pill>);

    await user.click(view.getByText('Pill'));

    expect(onClickMock).toBeCalled();
  });
});
