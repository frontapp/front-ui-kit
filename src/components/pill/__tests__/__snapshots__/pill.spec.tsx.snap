// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pill /> Spec Snapshots should render selected state 1`] = `
.c1 {
  grid-area: content;
}

.c0 {
  padding: 3px 6px;
  border-radius: 100px;
  border-radius: 4px;
  display: inline-grid;
  grid-template-areas: 'left-content content right-content';
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  box-sizing: border-box;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 16px;
  color: #ffffff;
  background-color: #77787a;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Pill
    </div>
  </div>
</div>
`;

exports[`<Pill /> Spec Snapshots should render using defaults 1`] = `
.c1 {
  grid-area: content;
}

.c0 {
  padding: 3px 6px;
  border-radius: 100px;
  border-radius: 4px;
  display: inline-grid;
  grid-template-areas: 'left-content content right-content';
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  box-sizing: border-box;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 16px;
  color: #19191a;
  background-color: rgba(0,0,0,0.1);
}

.c0:hover {
  color: #19191a;
  background-color: rgba(0,0,0,0.15);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Pill
    </div>
  </div>
</div>
`;

exports[`<Pill /> Spec Snapshots should support rendering a pill with icon 1`] = `
.c2 {
  width: 16px;
  height: 16px;
}

.c3 {
  grid-area: content;
}

.c0 {
  padding: 3px 6px;
  border-radius: 100px;
  border-radius: 4px;
  display: inline-grid;
  grid-template-areas: 'left-content content right-content';
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  box-sizing: border-box;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 16px;
  color: #19191a;
  background-color: rgba(0,0,0,0.1);
}

.c0:hover {
  color: #19191a;
  background-color: rgba(0,0,0,0.15);
}

.c1 {
  grid-area: left-content;
  margin-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          color=""
          height="16"
          name="Checkmark"
          preserveaspectratio="none"
          width="16"
        />
      </div>
    </div>
    <div
      class="c3"
    >
      Pill
    </div>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          color=""
          height="16"
          name="Checkmark"
          preserveaspectratio="none"
          width="16"
        />
      </div>
    </div>
  </div>
</div>
`;
